content/languages/python/authoring/index.md

```diff
@@ -0,0 +1,252 @@
+
+### Calling assertions
+
+The Python testing framework provides a set of useful [assertions](/languages/python/codewars-test/#assertions), but when used incorrectly, they can cause a series of problems:
+- Stacktraces of a crashing user solution can reveal details that should not be visible,
+- Use of an assertion not suitable for the given case may lead to incorrect test results,
+- Incorrectly used assertions may produce confusing or unhelpful messages.
+
+To avoid the above problems, calls to assertion functions should respect the following rules:
+- The expected value should be calculated _before_ invoking an assertion. The `expected` parameter passed to the assertion should not be a function call expression, but a value calculated directly beforehand.
+- Appropriate assertion functions should be used for each given test. `assert_equals` is not suitable in all situations. Use `assert_approx_equals` for floating point comparisons, `expect` for tests on boolean values, `expect_error` to test error handling.
+- Some additional attention should be paid to the order of parameters passed to assertion functions. It differs between various assertion libraries, and it happens to be quite often confused by authors, mixing up `actual` and `expected` in assertion messages. For the Python testing framework, the order is `(actual, expected)`.
+- One somewhat distinctive feature of Python assertions is that by default, a failed assertion does not cause a test case to fail early. It can lead to unexpected crashes when an actual value had already been asserted to be invalid, but the execution of the current test case was not stopped and following assertions continue to refer to it. This behavior can be overridden by passing the `allow_raise=True` argument to the assertion functions that support it.
+- To avoid unexpected crashes in tests, it's recommended to perform some additional assertions before assuming that the answer returned by the user solution has some particular type, form, or value. For example, if the test suite sorts the returned list to verify its correctness, an explicit assertion should be added to check whether the returned object is actually a list, and not, for example, `None`.
+
+
+## Additional restrictions
+
+### Accessing the solution file
+
+Some kata (for example, code-golf challenges or anti-cheat tests) would like to access and read the user solution file as text. It's possible and can be done by reading a file located at `/workspace/solution.txt`.
+
+### Blocking modules
+
+_TBD_
+
+
+## Example test suite
```

Comment:
might be cool to add a seciton in the example where the `it` blocks are built on the fly, inside a loop. That's a commom stuff, when you begin to build more complex test suites.

?

Reply:
I was thinking about it, did not get any good idea for this. But yes, it wold be worth adding 👍

Reply:
I wrote some stuff lik that, I can grab examples...

Reply:
out of the blue examples:

```code
    # loooooong data tests definitions, for TESTS_THE_TOKENIZER1, TESTS_THE_TOKENIZER2, ...
    # and defining the run_tests function that is using a loop to trun all the tests fro each section

    TEST_CONFIG = ( ('should correctly tokenize valid RSU program', TESTS_THE_TOKENIZER1),
                    ('should throw an error if one or more invalid tokens are detected', TESTS_THE_TOKENIZER2),
                    ('should correctly tokenize invalid RSU programs containing only valid tokens', TESTS_THE_TOKENIZER3),
                    ('Additional tests', NEW_TESTS_TOKENIZER),
                   )
    
    for msg,data in TEST_CONFIG:
        @test.it(msg)
        def _():
            run_tests(data)
```

here is the run_tests function, in case you find the intent clearer (not sure it will, because I have two kind of tests in there, changing what is tested depending on `exp` being `None` or not...):

```code
    def run_tests(config_tests):
        for program,exp,msg in config_tests:
            if exp:
                test.assert_equals(..., exp)
            else:
                test.expect_error(msg, lambda: ... )
```

Reply:
I have that one too, but using the timeout thing. Not sure it's a good idea to put it here.

```python

            CONFIG = [
                ((9,9), (17,17), circle(9,9,8)),
                ((0,0), (7,7), ((5,6),(5,8),(6,5),(6,9),(9,6),(8,5),(8,9),(9,8))),
                ((0,0), (9,9), circle(9,9,8)),
            ]
            for s,e,o in CONFIG:
                display(s,e,o)
                @test.timeout(.2)
                def _():
                    ret = _user_func(s,e,o)
                    Test.expect(ret is None,"Unreachable, the result should be None but was: "+str(ret))
```

Reply:
or with more levels:

```python

    msgIts = ["long stuff here", ...]
    for i,(S,E,exp,o) in enumerate([ ((0,14),  (0,35), 31, O),
                                     ((0,-20), (0,20), 24, O),
                                     ((0,-20), (0,35), 35, C1),
                                     ((0,-20), (0,35), 29, C2),
                                   ], 1):
        @Test.describe(f'Case #{i}')
        def _():
        
            timings=[]
            for j,(s,e) in enumerate([(S,E),(E,S)]):
                
                @Test.it(msgIts[j])
                def _():
                    display(s,e,o,limits)
                    act = _user_func(s,e,o)
                    Test.assert_equals(act, exp)
                
            @Test.it("All of that without losing in performances")
            def _():
                limit,a,b = 0.07, *timings
                # ...
                Test.expect(p<limit, f"Variation of execution time should be less than {limit:.1%}")
```

Reply:
I guess the first example is the most appropriate one...

Reply:
Added small example of dynamically generated test cases.